src/render/graph.rs

```diff
@@ -308,34 +314,43 @@ impl Graph {
                 // audio params are connected to the 'hidden' usize::MAX output, ignore them here
                 .filter(|edge| edge.other_index != usize::MAX)
                 .for_each(|edge| {
-                    let output_node = nodes.get_mut(&edge.other_id).unwrap();
+                    let mut output_node = nodes.get(&edge.other_id).unwrap().borrow_mut();
                     output_node.has_inputs_connected = true;
                     let signal = &node.outputs[edge.self_index];
-                    let channel_config = &output_node.channel_config;
+                    let channel_config = &output_node.channel_config.clone();
 
                     output_node.inputs[edge.other_index].add(signal, channel_config);
                 });
 
-            // Check if we can decommission this node (end of life)
-            if node.can_free(tail_time) {
-                // Node is dropped, we should perform a new topological sort of the audio graph
-                nodes_dropped = true;
+            let can_free = node.can_free(tail_time);
 
-                // Nodes are only dropped when they do not have incoming connections.
-                // But they may have AudioParams feeding into them, these can de dropped too.
-                nodes.retain(|_id, n| !n.outgoing_edges.iter().any(|e| e.other_id == *index));
-            } else {
-                // Node is not dropped.
+            // Node is not dropped.
+            if !can_free {
                 // Reset input buffers as they will be summed up in the next render quantum.
                 node.inputs
                     .iter_mut()
                     .for_each(AudioRenderQuantum::make_silent);
 
                 // Reset input state
                 node.has_inputs_connected = false;
+            }
 
-                // Re-insert node in graph
-                nodes.insert(*index, node);
+            drop(node); // release borrow of self.nodes
+
+            // Check if we can decommission this node (end of life)
+            if can_free {
+                // Node is dropped, we should perform a new topological sort of the audio graph
```

Comment:
Just one last thing maybe, this comment is confusing because we actually don't trigger a new sort of the graph here